### DISCUSSION TITLE
cue vet not failing a schema validation

### DISCUSSION THREAD

**Discussion by philippgille** [cue-lang/cue#2424] [Q&A] [answered]
Hello :wave: , I'm playing around with CUE, mostly looking into the schema validation. There is one instance where I expected the validation to fail, but it succeeds. I'm probably doing something wrong, can you help me find my mistake please?

The schema is this, where I want the target document be a list of `Foo`:

```cue
#Foo: {
  name: string
  bars: [...#Bar]
}

#Bar: {
  name: string
  bla: [...#Baz]
}

#Baz: {
  note: string
}

[...#Foo]
```

And the YAML is this:

```yaml
- name: foo
  bars:
  - name: bar
    bla:
    - note: xyz
```

Running `cue vet data.yml schema.cue` works fine as expected.

Then I decrease the indentation of some fields:

```yaml
- name: foo
  bars:
  - name: bar
  bla:
  - note: xyz
```

Running `cue vet data.yml schema.cue` succeeds, but shouldn't this be invalid based on the schema? `bla` is not a field of `Foo`, but of `Bar`.

If `Foo` were "open", i.e. with `...` at the end of the struct, I would understand. But it's not open, right?

I also tried the other way around to see if my YAML is bad, with `cue import`, to see if the results ended up the same (which would explain why the schema validation works on both), but they're different:

The one I expect to be valid:

```
$ cue import -R -s -o - data.yml 
[{
	name: "foo"
	bars: [{
		name: "bar"
		bla: [{
			note: "xyz"
		}]
	}]
}]
```

The one I expect to be invalid:

```
$ cue import -R -s -o - data2.yml 
[{
	name: "foo"
	bars: [{
		name: "bar"
	}]
	bla: [{
		note: "xyz"
	}]
}]
```

---

Previous similar discussions I found but don't seem directly related:

- https://github.com/cue-lang/cue/discussions/1949
- https://github.com/cue-lang/cue/discussions/1777
- https://github.com/cue-lang/cue/discussions/876

**Comment by opax** [comment, marked as answer]
Top-level lists are indeed a little tricky. See [here](https://github.com/cue-lang/cue/discussions/1771?sort=top#discussioncomment-3032547) for some more details.

**Reply by philippgille** [reply]
Thanks, that helped :+1: 

TL;DR for others reading this:
When you have a top level list `[...#Foo]`, then the data can add new fields to Foo instances, even though the struct is supposed to be "closed".
The workaround is to define another top level element, and then if you don't want to add that to your data you're vetting (e.g. the YAML should have the list as top level, not a useless single field) you can use flags to tell CUE about it when vetting.
Two concrete examples:
1. Use `schema: [...#Foo]` in the CUE file, then `cue vet -l '"schema"' data.yml schema.cue` (double quotes inside single quotes)
2. Use `#Schema: [...#Foo]` in the CUE file, then `cue vet -d '#Schema' data.yml schema.cue` (can also be lowercase)
